src/Formats/ProtobufSerializer.cpp

```diff
@@ -3353,20 +3353,63 @@ namespace
+                                if(nested_data_type->getTypeId() == TypeIndex::Array)
+                                    nested_some_arrays = true ;
+                                else
+                                    nested_all_arrays = false ;
+                            }
+                        }
+
+                        /// Since all selected nested columns are of type array, bypass exception flow :
+                        /// Attempt serialization of nested columns as an array of nested messages directly
+                        if (nested_all_arrays)
+                        {
+                            for (DataTypePtr & dt : nested_data_types)
+                                dt = assert_cast<const DataTypeArray &>(*dt).getNestedType();
+
+                            std::vector<size_t> used_column_indices_in_nested;
+                            auto nested_message_serializer = buildMessageSerializerImpl(
+                                nested_column_names.size(),
+                                nested_column_names.data(),
+                                nested_data_types.data(),
+                                *field_descriptor->message_type(),
+                                /* with_length_delimiter = */ false,
+                                google_wrappers_special_treatment,
+                                field_descriptor,
+                                used_column_indices_in_nested,
+                                /* columns_are_reordered_outside = */ true,
+                                /* check_nested_while_filling_missing_columns = */ false);
+
+                            /// `columns_are_reordered_outside` is true because column indices are
+                            /// going to be transformed and then written to the outer message,
+                            /// see add_field_serializer() below.
+
+                            if (nested_message_serializer)
+                            {
+                                std::vector<std::string_view> column_names_used;
+                                column_names_used.reserve(used_column_indices_in_nested.size());
+                                for (size_t i : used_column_indices_in_nested)
+                                    column_names_used.emplace_back(nested_column_names[i]);
+                                auto field_serializer = std::make_unique<ProtobufSerializerFlattenedNestedAsArrayOfNestedMessages>(
+                                    std::move(column_names_used), field_descriptor, std::move(nested_message_serializer), get_root_desc_function);
+                                transformColumnIndices(used_column_indices_in_nested, nested_column_indices);
+                                add_field_serializer(column_name, std::move(used_column_indices_in_nested), *field_descriptor, std::move(field_serializer));
+                                break;
+                            }
+
+                            continue ;
```

Comment:
Is this `continue` needed if there's already a `break` above?

Reply:
Actually I think it's wrong, it will just not serialize if `nested_message_serializer` is a `nullptr`

Reply:
> Is this continue needed if there's already a break above?

I want to say yes, because the inner loop is for proto fields, and the outer loop is for columns ; 
the original code appears to be written in such a way as to maximize matching opportunities,
and check further down if any columns/fields have not been matched

I'm happy to discuss this in more details, I agree it may be fairly confusing 

however after your suggestion for factorizing into lambdas I've rewritten the control flow such that this continue is not necessary anymore (hopefully I haven't made a mistake) :pray: